CONTRIBUTING.md

```diff
@@ -24,19 +24,153 @@ Please follow these steps when making pull requests (PRs):
     Please give the PR a descriptive name.
 
 1.  As you make progress on your PR, keep your branch up-to-date with the `master` branch which
-        may have been updated after starting your PR.
+        may have been updated *after* starting your PR.
     Your PR **MUST** be updated to reflect changes to `master` in order to be merged.
-    Use the following procedure for updating your branch:
+    Use the following procedure for updating your branch and are ready to commit your changes:
 
-    <!--- language: lang-shell -->
+    <!--- language: lang-none -->
 
-        # While on your PR branch <issue-branch>
-        git commit
+        ## While on your PR branch <issue-branch> hosted at <your-remote> repository:
+        git commit # or git stash                                               ## Line A
         git checkout master
-        git pull origin --rebase
+
+        git pull <illixr-remote> master --rebase && git fetch <illixr-remote>   ## Line B
+
         git checkout <issue-branch>
-        git rebase master
-        git push origin <issue-branch> --force-with-lease
+        git rebase master                                                       ## Line C
+
+        ## If you stashed your changes on 'Line A':
+        git stash apply <stash-number> && git commit
+
+        git push <your-remote> <issue-branch> --force-with-lease                ## Line D
+
+    For ILLIXR team members (others jump [here][12]):
+
+    -   In the example above, `<illixr-remote>` and `<your-remote>` are the same.
+
+    -   When collaborating on branches in our repository, `Line B` may pull in changes that overwrite
+            the git commit history when performing `Line C`.
+        Subsequently, performing `Line D` will rewrite the history in the public branch.
+        To preserve branch commit histories in the case that a rollback is needed, we will employ
+            a checkpointing process for force updated branches.
+        This process will be manually performed, but may be automated in the future.
+
+        If `Line B` shows an update to master, the following example illustrates your local repository
+            just after performing `Line B`:
+
+        <!--- language: lang-none -->
+
+            A -- B -- C -- P -- Q -- R                                          ## master
+                       \
+                        D -- E -- F                                             ## issue-123-fixing-bug
+
+        In this example, commits `P`, `Q`, and `R` have been merged to `master`
+            (from feature branches not shown) after feature branch `issue-123-fixing-bug` was
+            forked from `master`.
+
+        To checkpoint the `issue-123-fixing-bug` while it is checked out:
```

Comment:
Most of the "complexity" in this document is coming from checkpointing. Why is checkpointing necessary, does every developer need to do it, and can we live without it?

Reply:
> Why is checkpointing necessary,

The need to checkpoint arises from using rebases instead of merges. Since rebases rewrite history, we have no choice but to do force-updates when updating a feature branch with commits pushed to `master`. Force-updates can cause lossy git fetches if care is not taken (as described in the document). Checkpoints are only relevant for public branches that are updated via rebases (currently proposed to be all feature branches), _and_ are being contributed to my more than one developer.

> does every developer need to do it

Checkpoints can only be done for branches/repos for which we have write permissions. Thus, there is no need to consider checkpointing for external repos from which a PR is generated. For ILLIXR-owned repos, branches that are being worked on (i.e. actively pushing commits) by multiple developers will require checkpointing. This means that checkpointing is primarily an ILLIXR-team developer concern.

> can we live without it

We can go back to the usual merge based flow. However, we lose some of the benefits afforded by rebased linear branch histories. This is also a perfectly fine solution, but maintenance up-front will be traded off for maintenance on-demand for the maintainer navigating merged histories. Although checkpointing as described is quite heavyweight for a manual process, I think this system would work quite well with automation. Checkpointing is needed to resolve collaboration conflicts, but conflicts do not frequently occur in our workflow currently. Hopefully, this means that we can enforce this manually without the overhead of handling conflicts often. Once automated, we won't have to worry about this any longer.